File: frontend/src/api/activity.ts
import api from './client'

export interface Activity {
  id: number
  name: string
  activity_type: string
  description?: string
  location?: string
  activity_date: string
  start_time: string
  end_time: string
  max_participants?: number
  instructor?: string
  equipment_needed?: string
  risk_level: string
  status: string
  created_by?: number
  created_at: string
  updated_at: string
}

export interface ActivitySignIn {
  id: number
  activity_id: number
  elder_id: number
  sign_in_time?: string
  sign_out_time?: string
  sign_in_by_id?: number
  participation_status: string
  health_before?: string
  health_after?: string
  performance_rating?: number
  remark?: string
  created_at: string
  updated_at: string
}

export const activityApi = {
  getList: (params?: {
    page?: number
    page_size?: number
    activity_type?: string
    status?: string
    keyword?: string
  }) => {
    return api.get<any>('/activities', { params })
  },

  getDetail: (id: number) => {
    return api.get<Activity>(`/activities/${id}`)
  },

  create: (data: Partial<Activity>) => {
    return api.post<Activity>('/activities', data)
  },

  update: (id: number, data: Partial<Activity>) => {
    return api.put<Activity>(`/activities/${id}`, data)
  },

  remove: (id: number) => {
    return api.delete(`/activities/${id}`)
  },

  signIn: (activityId: number, data: { elder_id: number; health_before?: string }) => {
    return api.post<ActivitySignIn>(`/activities/${activityId}/sign-in`, data)
  },

  getSignIns: (activityId: number) => {
    return api.get<ActivitySignIn[]>(`/activities/${activityId}/sign-ins`)
  },

  updateSignIn: (id: number, data: Partial<ActivitySignIn>) => {
    return api.patch<ActivitySignIn>(`/activity-sign-ins/${id}`, data)
  },
}

File: backend/app/models/activity.py
from datetime import datetime, date
from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, Text, Date, Time, Boolean
from sqlalchemy.orm import relationship

from app.database import Base


class Activity(Base):
    __tablename__ = "activities"

    id = Column(Integer, primary_key=True, index=True)
    name = Column(String(200), nullable=False)
    activity_type = Column(String(50), nullable=False)
    description = Column(Text, nullable=True)
    location = Column(String(200), nullable=True)
    activity_date = Column(Date, nullable=False)
    start_time = Column(Time, nullable=False)
    end_time = Column(Time, nullable=False)
    max_participants = Column(Integer, nullable=True)
    instructor = Column(String(100), nullable=True)
    equipment_needed = Column(Text, nullable=True)
    risk_level = Column(String(20), nullable=False, default="low")
    status = Column(String(20), nullable=False, default="scheduled")
    created_by = Column(Integer, ForeignKey("users.id"), nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    sign_ins = relationship("ActivitySignIn", back_populates="activity", cascade="all, delete-orphan")


class ActivitySignIn(Base):
    __tablename__ = "activity_sign_ins"

    id = Column(Integer, primary_key=True, index=True)
    activity_id = Column(Integer, ForeignKey("activities.id"), nullable=False, index=True)
    elder_id = Column(Integer, ForeignKey("elders.id"), nullable=False, index=True)
    sign_in_time = Column(DateTime, nullable=True)
    sign_out_time = Column(DateTime, nullable=True)
    sign_in_by_id = Column(Integer, ForeignKey("users.id"), nullable=True)
    participation_status = Column(String(20), nullable=False, default="signed_in")
    health_before = Column(String(200), nullable=True)
    health_after = Column(String(200), nullable=True)
    performance_rating = Column(Integer, nullable=True)
    remark = Column(Text, nullable=True)
    created_at = Column(DateTime, default=datetime.utcnow)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow)

    activity = relationship("Activity", back_populates="sign_ins")
    elder = relationship("Elder", back_populates="activity_sign_ins")
    sign_in_by = relationship("User", back_populates="activity_sign_ins")
